### ISSUE TITLE
UI reports agent as "disabled (building)", API only reports "disabled"

### ISSUE THREAD

**Issue by grahamc** [gocd/gocd#1438]
Here is what the API returns for my agent, where I was expecting the status to demonstrate somehow that this agent was building:

```
{
  "_links": {
    "self": {
      "href": "https:/.../go/api/agents/1e8be59d-471f-4d27-b4ca-ff73d5f9fb8e"
    },
    "doc": {
      "href": "http://api.go.cd/#agents"
    },
    "find": {
      "href": "https://.../go/api/agents/:uuid"
    }
  },
  "uuid": "1e8be59d-471f-4d27-b4ca-ff73d5f9fb8e",
  "hostname": "b2fabf99838e",
  "ip_address": "172.17.0.156",
  "enabled": false,
  "sandbox": "/var/lib/go-agent",
  "status": "Disabled",
  "operating_system": "Linux",
  "free_space": 274511863808,
  "resources": [

  ],
  "environments": [
    "Build"
  ]
}
```

Is there another way to determine if an agent is doing work?

**Comment by grahamc** [comment]
It looks like the logic represented here: https://github.com/gocd/gocd/blob/f52a72f9b7c23919470dbdbc47139ef1dc95acf8/server/webapp/WEB-INF/rails.new/app/helpers/agents_helper.rb#L41-L59 might need to be backported to https://github.com/gocd/gocd/blob/f52a72f9b7c23919470dbdbc47139ef1dc95acf8/server/src/com/thoughtworks/go/server/ui/AgentViewModel.java#L90-L92, as the helper is used in the HTML view and the view model is used exclusively (in the presenter) for the API. The logic here https://github.com/gocd/gocd/blob/f52a72f9b7c23919470dbdbc47139ef1dc95acf8/common/src/com/thoughtworks/go/domain/AgentInstance.java#L128-L132 short-circuits if the agent is Disabled, preventing a `disabled (building)` status.

It might be more correct to have an enabled / disabled field, and then a status field which is independent of enabled / disabled. IE, an agent could be `disabled` AND `idle` in the UI.

**Comment by grahamc** [comment]
*bump* I'm using a script to disable all agents before a configuration update. Because I can't tell if a build is still happening, it could potentially cause problems:

```python
#!/usr/bin/env python3

import requests
from pprint import pprint
import time
import json

session = requests.Session()
session.headers.update({'Accept': 'application/vnd.go.cd.v1+json'})

base_url = 'https://go'

def all_agents():
    return session.get('%s/go/api/agents' % base_url).json()['_embedded']['agents']

def should_disable_agent(agent):
    return agent['environments'] != ['goserver']

def disable_agent(agent):
    r = session.patch(
            agent['_links']['self']['href'],
            data=json.dumps({"enabled": False}),
            headers={"Content-Type": "application/json"}
            )

def any_agent_has_work(agents):
    for agent in agents:
        if agent_has_work(agent):
            return agent

    return False

def agent_has_work(agent):
    pprint(agent)
    return False

agents = [agent for agent in all_agents() if should_disable_agent(agent)]
for agent in agents:
    disable_agent(agent)


while True:
    with_work = any_agent_has_work(agents)
    if with_work:
        print("Agent %s has work ... trying again in 5 seconds" % with_work['uuid'])
        time.sleep(1)
        continue
    break
```

Note that `agent_has_work` isn't possible to implement at the moment.

**Comment by grahamc** [comment]
From Gitter: 

```
  ketan | I’m looking at #1438, am I right to paraphrase that `status` should not contain `Disabled`
        | since there’s already a `enabled` field in the JSON?
grahamc | I think that is probably right
  ketan | we’re trying to make sure things are more consistent going forward, which means that there
        | may be specific cases where the view and APIs may differ in what information they show and
        | how they show it.
  ketan | cool
grahamc | but the real core of it is there should be an indication that the agent is building
  ketan | gotcha, it seems like `disabled` trumps `building`
grahamc | that is the one I care about specifically, a way to programatically know that an agent is
        | working. and yeah, it totally does
  ketan | thanks for the confirmation and the pointers on the issue, I was able to reproduce the issue
        | and I’ll fix it
```

**Comment by arvindsv** [comment]
@ketan: Shouldn't this be closed? Need to update docs for 15.3, I suppose.

**Comment by rajiesh** [comment]
verified this v2 of Agent API with more status information tested on 15.3.0-2672(47fd7343016e4a9cda3f8a871539007c2d1cf1b9) and working as expected
